components/dashboard/src/whatsnew/WhatsNew.tsx

```diff
@@ -0,0 +1,100 @@
+import { UserContext } from "../user-context";
+import { useContext, useState } from "react";
+import { getGitpodService } from "../service/service";
+
+const allEntries: WhatsNewEntry[] = [
+    WhatsNewEntry202106,
+    WhatsNewEntry202104,
+]
+
+export const shouldSeeWhatsNew = (user: User, news: { newsKey: string, maxUserCreationDate: string }[] = allEntries) => {
+    const whatsNewSeen = user?.additionalData?.whatsNewSeen;
+    return news.some(x => user.creationDate <= x.maxUserCreationDate && (!whatsNewSeen || !whatsNewSeen[x.newsKey]));
+}
+
+export function WhatsNew(props: { onClose: () => void }) {
+    const { user, setUser } = useContext(UserContext);
+
+    const _unseenEntries = allEntries.filter(x => user && shouldSeeWhatsNew(user, [x])) || [];
+    const [visibleEntry, setVisibleEntry] = useState(_unseenEntries.pop());
+    const [unseenEntries, setUnseenEntries] = useState(_unseenEntries);
+
+    const markAsSeen = async (user?: User, ...news: (WhatsNewEntry | undefined)[]) => {
+        if (!user) {
+            return;
+        }
+
+        for (const n of news.filter(x => x && x.actionAfterSeen)) {
+            user = await n!.actionAfterSeen!(user);
+        };
+
+        const additionalData = user.additionalData = user.additionalData || {};
+        additionalData.whatsNewSeen = additionalData.whatsNewSeen || {};
+        const now = new Date().toISOString();
+        for (const newsKey of (news.filter(x => x !== undefined) as WhatsNewEntry[]).map(x => x.newsKey)) {
+            additionalData.whatsNewSeen[newsKey] = now;
+        }
+        user = await getGitpodService().server.updateLoggedInUser({
+            additionalData
+        });
+        setUser(user);
+    };
+
+    const internalClose = async () => {
+        await markAsSeen(user, ...unseenEntries, visibleEntry);
+        props.onClose();
+    };
+
+    const hasNext = () => unseenEntries.length > 0;
+
+    const next = async () => {
+        if (unseenEntries.length === 0) {
+            return;
+        }
+        visibleEntry && await markAsSeen(user, visibleEntry);
+        const _unseenEntries = unseenEntries;
+        setVisibleEntry(_unseenEntries.pop());
+        setUnseenEntries(_unseenEntries);
+    };
+
+    return <Modal
+        visible={!!visibleEntry}
+        onClose={internalClose}
+    >
+        <h3 className="pb-4">What's New 🎁</h3>
+        <>{visibleEntry && user ? visibleEntry.children(user, setUser) : <></>}</>
+        {
+            hasNext() ?
+                <>
+                    <div className="flex items-start justify-end mt-6 space-x-2">
+                        <button className="ml-2 mt-0 secondary" onClick={internalClose}>Dismiss All</button>
+                        <div className="flex flex-col items-center">
+                            <button className="ml-2" onClick={next}>Next</button>
+                            <div className="text-xs italic mt-1 ml-2 -mb-2">({unseenEntries.length} left)</div>
```

Comment:
**suggestion**: Alternatively, we could left align this  on the same row with the button and read 1 of 2.

Reply:
Yes, I'm actually not that happy with the text under the “Next” button either. However, I personally don't like the schema “1 of 2” here because
1. there is no button back therefore the number of already seen pages seems not that important to me and
2. technically, I don't have the number of the total pages because the total number changes (decreases) at that moment I click on “Next“.

But we can address this in a follow-up PR if you like.

I would propose that change:

```diff
--- a/components/dashboard/src/whatsnew/WhatsNew.tsx
+++ b/components/dashboard/src/whatsnew/WhatsNew.tsx
@@ -76,12 +76,10 @@ export function WhatsNew(props: { onClose: () => void }) {
         {
             hasNext() ?
                 <>
-                    <div className="flex items-start justify-end mt-6 space-x-2">
-                        <button className="ml-2 mt-0 secondary" onClick={internalClose}>Dismiss All</button>
-                        <div className="flex flex-col items-center">
-                            <button className="ml-2" onClick={next}>Next</button>
-                            <div className="text-xs italic mt-1 ml-2 -mb-2">({unseenEntries.length} left)</div>
-                        </div>
+                    <div className="flex items-center justify-end mt-6 space-x-2">
+                        <div className="text-sm mr-auto italic">{unseenEntries.length} page{unseenEntries.length > 1 ? "s" : ""} left</div>
+                        <button className="ml-2 secondary" onClick={internalClose}>Dismiss All</button>
+                        <button className="ml-2" onClick={next}>Next</button>
                     </div>
                 </>
```

| Before        | After           |
| ------------- |-------------|
| ![image](https://user-images.githubusercontent.com/24960040/123469689-a8c23080-d5f3-11eb-8082-ce973e0b6dab.png)      |  ![image](https://user-images.githubusercontent.com/24960040/123470321-8977d300-d5f4-11eb-9c60-b7860a5780ab.png) |



![image](https://user-images.githubusercontent.com/24960040/123470370-998fb280-d5f4-11eb-96a1-85b915cdafe2.png)

What do you think?

Reply:
LET'S DO IT. Visually it's much better. I would only rephrase _1 page left_ to _1 more update_ or _1 item left_ to move away from the page reference as this could be confusing. Your call! 📞